src/common/src/array/list_array.rs

```diff
@@ -429,22 +449,19 @@ impl<'a> ListRef<'a> {
         &self,
         serializer: &mut memcomparable::Serializer<impl BufMut>,
     ) -> memcomparable::Result<()> {
+        let mut inner_serializer = memcomparable::Serializer::new(vec![]);
         match self {
-            ListRef::Indexed { arr, idx } => {
-                let offsets = arr.offsets[*idx]..arr.offsets[*idx + 1];
-                serializer.serialize_u32(offsets.len() as u32)?;
-                offsets
-                    .map(|o| arr.value.value_at(o))
-                    .try_for_each(|datum_ref| serialize_datum_ref_into(&datum_ref, serializer))?
-            }
-            ListRef::ValueRef { val } => {
-                serializer.serialize_u32(val.values.len() as u32)?;
-                val.values
-                    .iter()
-                    .try_for_each(|datum| serialize_datum_into(datum, serializer))?
-            }
+            ListRef::Indexed { arr, idx } => (arr.offsets[*idx]..arr.offsets[*idx + 1])
+                .map(|o| arr.value.value_at(o))
+                .try_for_each(|datum_ref| {
+                    serialize_datum_ref_into(&datum_ref, &mut inner_serializer)
+                })?,
```

Comment:
This might be wrong if we compare the 2 rows (have not tested on this branch yet):
`values (array[1, 2], 3), (array[1], 3) order by 1, 2;`

The memcomparable encoding for varchar contains a signal byte after each chunk to indicate termination, and prevents accidental comparison between a longer value (`array[1, 2]`) to the next column of shorter value (`3` after `array[1]`).

https://github.com/singularity-data/risingwave/blob/dfefd0e4cfdf0a4507e4ff8d7d286ba563186d42/src/utils/memcomparable/src/ser.rs#L201-L205

Similar test case in varchar:
`values ('ab', ''), ('a', 'c') order by 1, 2;`

Applying to list type here, it should also have a signal byte after every N elements, with 0 padding for the last chunk. The choice of a proper N can be discussed separately.

Reply:
tests added